src/runtime/rpc/rpc_server.cc

```diff
@@ -0,0 +1,317 @@
+ * \param sock The socket information
+ * \param addr The socket address information
+ */
+void ServerLoopProc(common::TCPSocket sock, common::SockAddr addr) {
+    // Server loop
+    auto env = RPCEnv();
+    RPCServerLoop(sock.sockfd);
+    LOG(INFO) << "Finish serving " << addr.AsString();
+    env.Remove();
+}
+
+/*!
+ * \brief GetTimeOutFromOpts Parse and get the timeout option.
+ * \param opts The option string
+ * \param timeout value after parsing.
+ */
+int GetTimeOutFromOpts(std::string opts) {
+  std::string cmd;
+  std::string option = "--timeout=";
+
+  if (opts.find(option) == 0) {
+    cmd = opts.substr(opts.find_last_of(option) + 1);
+    CHECK(IsNumber(cmd)) << "Timeout is not valid";
+    return std::stoi(cmd);
+  }
+  return 0;
+}
+
+/*!
+ * \brief ListenLoopProc The listen process.
+ * \param sock The socket information
+ * \param port The port of the RPC
+ * \param tracker The address of RPC tracker in host:port format.
+ * \param key The key used to identify the device type in tracker
+ * \param custom_addr Custom IP Address to Report to RPC Tracker.
+ */
+void ListenLoopProc(common::TCPSocket sock,
+                    int port,
+                    std::string tracker_addr,
+                    std::string rpc_key,
+                    std::string custom_addr) {
+  common::TCPSocket tracker_conn;
+  while (1) {
+    common::TCPSocket conn;
+    common::SockAddr addr("0.0.0.0", 0);
+    std::string opts;
+    try {
+      // step 1: setup tracker and report to tracker
+      if (!tracker_addr.empty() && (tracker_conn.IsClosed())) {
+        tracker_conn = ConnectWithRetry(tracker_addr);
+
+        int code = kRPCTrackerMagic;
+        CHECK_EQ(tracker_conn.SendAll(&code, sizeof(code)), sizeof(code));
+        CHECK_EQ(tracker_conn.RecvAll(&code, sizeof(code)), sizeof(code));
+        CHECK_EQ(code, kRPCTrackerMagic) << tracker_addr.c_str() << " is not RPC Tracker";
+
+        std::ostringstream ss;
+        ss << "[" << static_cast<int>(TrackerCode::kUpdateInfo)
+           << ", {\"key\": \"server:"<< rpc_key << "\"}]";
+        SendData(tracker_conn, ss.str());
+
+        // Receive status and validate
+        std::string remote_status = RecvData(tracker_conn);
+        CHECK_EQ(std::stoi(remote_status), static_cast<int>(TrackerCode::kSuccess));
+      }
+      // step 2: wait for in-coming connections
+      AcceptConnection(sock, port, tracker_conn, rpc_key, custom_addr, &conn, &addr, &opts);
+    }
+    catch (const char* msg) {  // (socket.error, IOError):
+        // retry when tracker is dropped
+        if (!tracker_conn.IsClosed()) {
+          tracker_conn.Close();
+        }
+        continue;
+    }
+    catch (std::exception& e) {
+          // Other errors
+          LOG(FATAL) << "Exception standard: " << e.what();
+    }
+
+    #if defined(__linux__) || defined(__ANDROID__)
+      // step 3: serving
+      int timeout = GetTimeOutFromOpts(opts);
+      auto pid = fork();
```

Comment:
The fork policy won't really work on the android, consider use the suicide strategy